File: app/src/main/java/com/th/forge/taxiorders/OrderDetailFragment.java
package com.th.forge.taxiorders;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.th.forge.taxiorders.entity.Order;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.io.Serializable;
import java.text.NumberFormat;
import java.text.SimpleDateFormat;
import java.util.Currency;

import okhttp3.ResponseBody;
import retrofit2.Response;

public class OrderDetailFragment extends Fragment {
    private static final String ORDER_KEY = "order_key";

    private OnFragmentInteractionListener interactionListener;
    private Order order;
    private ImageView imageView;

    public interface OnFragmentInteractionListener {
        void onFragmentInteraction(String title, String subtitle);
    }

    public static OrderDetailFragment newInstance(Serializable order) {
        Bundle args = new Bundle();
        args.putSerializable(ORDER_KEY, order);
        OrderDetailFragment fragment = new OrderDetailFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        interactionListener = (OnFragmentInteractionListener) context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            order = (Order) getArguments().getSerializable(ORDER_KEY);
        }
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_order_detail, container, false);
        imageView = view.findViewById(R.id.detail_image);
        TextView startStreet = view.findViewById(R.id.detail_start_address);
        TextView endStreet = view.findViewById(R.id.detail_end_address);
        TextView driverName = view.findViewById(R.id.detail_vehicle_driver);
        TextView vehicleModel = view.findViewById(R.id.detail_vehicle_model);
        TextView vehicleNumber = view.findViewById(R.id.detail_vehicle_reg_number);
        TextView priceField = view.findViewById(R.id.detail_price);
        updateTitle();
        startStreet.setText(order.getStartAddress().getAddress());
        endStreet.setText(order.getEndAddress().getAddress());
        driverName.setText(order.getVehicle().getDriverName());
        vehicleModel.setText(order.getVehicle().getModelName());
        vehicleNumber.setText(order.getVehicle().getRegNumber());
        Currency currency = Currency.getInstance(order.getPrice().getCurrency());
        NumberFormat formatter = NumberFormat.getInstance();
        formatter.setMinimumFractionDigits(2);
        formatter.setCurrency(currency);
        priceField.setText(String.format("%s %s", formatter.format(order.getPrice().getAmount()), getResources().getString(R.string.rubleSymbol)));
        new FetchImageTask().execute();
        return view;
    }

    private void updateTitle() {
        if (interactionListener != null) {
            String title = getString(R.string.detail_title_field)
                    + new SimpleDateFormat("dd.MM.yyyy").format(order.getOrderTime().getTime());
            String subtitle = getString(R.string.detail_subtitle_field)
                    + new SimpleDateFormat("HH:mm").format(order.getOrderTime().getTime());
            interactionListener.onFragmentInteraction(title, subtitle);
        }
    }

    private class FetchImageTask extends AsyncTask<Void, Void, Bitmap>{

        @Override
        protected Bitmap doInBackground(Void... voids) {
            return getImage();
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            imageView.setImageBitmap(bitmap);
        }
    }

    private Bitmap getImage() {
        String imagePath = order.getVehicle().getPhoto();
        File cachedImage = new File(getActivity().getFilesDir(), imagePath);
        Bitmap bmp;
        long currentTime = System.currentTimeMillis();
        if (cachedImage.exists() && (currentTime - cachedImage.lastModified() < 1 * 1000 * 60)) {
            bmp = BitmapFactory.decodeFile(cachedImage.getAbsolutePath());
        } else {
            cachedImage.delete();
            try {
                Response<ResponseBody> response = App.getApiService().getImage(imagePath).execute();
                if(response.body()!=null){
                    OutputStream os;
                    os = new FileOutputStream(cachedImage);
                    os.write(response.body().bytes());
                    os.flush();
                    os.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            bmp = BitmapFactory.decodeFile(cachedImage.getAbsolutePath());
        }
        return bmp;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        interactionListener = null;
    }
}

File: app/src/main/java/com/th/forge/taxiorders/OrderDetailFragment.kt
package com.th.forge.taxiorders

import android.content.Context
import android.graphics.Bitmap
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.th.forge.taxiorders.entity.Order
import com.th.forge.taxiorders.repo.ImageRepository
import com.th.forge.taxiorders.utils.CurrencyParser
import com.th.forge.taxiorders.utils.DateTimeParser

import java.io.Serializable


class OrderDetailFragment : Fragment() {

    private var interactionListener: OnFragmentInteractionListener? = null
    private var order: Order? = null
    private var image: Bitmap? = null

    /* ToDo ButterKnife */
    private var imageView: ImageView? = null
    private var startStreet: TextView? = null
    private var endStreet: TextView? = null
    private var driverName: TextView? = null
    private var vehicleModel: TextView? = null
    private var vehicleNumber: TextView? = null
    private var priceField: TextView? = null

    interface OnFragmentInteractionListener {
        fun onFragmentInteraction(title: String, subtitle: String)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        interactionListener = context as OnFragmentInteractionListener?
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            order = arguments!!.getSerializable(ORDER_KEY) as Order
        }
        if (savedInstanceState != null) {
            Log.d(LOG_TAG, "onCreateView, saved != null")
            image = savedInstanceState.getParcelable(SAVE_IMAGE)
        } else {
            Log.d(LOG_TAG, "onCreateView, saved = null")
            FetchImageTask(order!!.vehicle!!.photo).execute()
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_order_detail, container, false)
        initUi(view)
        setData()
        return view
    }

    private fun initUi(view: View) {
        updateTitle()
        imageView = view.findViewById(R.id.detail_image)
        startStreet = view.findViewById(R.id.detail_start_address)
        endStreet = view.findViewById(R.id.detail_end_address)
        driverName = view.findViewById(R.id.detail_vehicle_driver)
        vehicleModel = view.findViewById(R.id.detail_vehicle_model)
        vehicleNumber = view.findViewById(R.id.detail_vehicle_reg_number)
        priceField = view.findViewById(R.id.detail_price)
    }

    private fun updateTitle() {
        if (interactionListener != null) {
            val title = String.format("%s %s", getString(R.string.detail_title_field),
                    DateTimeParser.getReadableString("dd.MM.yyyy", order!!.orderTime!!.time))
            val subtitle = String.format("%s %s", getString(R.string.detail_subtitle_field),
                    DateTimeParser.getReadableString("HH:mm", order!!.orderTime!!.time))
            interactionListener!!.onFragmentInteraction(title, subtitle)
        }
    }

    private fun setData() {
        imageView!!.setImageBitmap(image)
        startStreet!!.text = order!!.startAddress!!.address
        endStreet!!.text = order!!.endAddress!!.address
        driverName!!.text = order!!.vehicle!!.driverName
        vehicleModel!!.text = order!!.vehicle!!.modelName
        vehicleNumber!!.text = order!!.vehicle!!.regNumber
        priceField!!.text = CurrencyParser
                .getFormattedPrice(order!!.price!!.amount,
                        order!!.price!!.currency, resources.getString(R.string.rubleSymbol))
    }

    private inner class FetchImageTask internal constructor(private val imagePath: String) : AsyncTask<Void, Void, Bitmap>() {

        override fun doInBackground(vararg voids: Void): Bitmap? {
            Log.d(LOG_TAG, "I'm HERE!!!")
            image = ImageRepository.getImage(activity!!.filesDir, imagePath)
            return image
        }

        override fun onPostExecute(bitmap: Bitmap) {
            imageView!!.setImageBitmap(bitmap)
        }

    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        Log.d(LOG_TAG, "onSaveInstance")
        if (image != null) {
            outState.putParcelable(SAVE_IMAGE, image)
        }
    }

    override fun onDetach() {
        super.onDetach()
        interactionListener = null
    }

    companion object {
        private val ORDER_KEY = "order_key"
        private val LOG_TAG = "DETAILFRGMNT"
        private val SAVE_IMAGE = "SAVE_IMAGE"

        fun newInstance(order: Serializable): OrderDetailFragment {
            val args = Bundle()
            args.putSerializable(ORDER_KEY, order)
            val fragment = OrderDetailFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
